src/nnue/layers/affine_transform_sparse_input.h

```diff
@@ -0,0 +1,255 @@
+/*
+  Stockfish, a UCI chess playing engine derived from Glaurung 2.1
+  Copyright (C) 2004-2023 The Stockfish developers (see AUTHORS file)
+
+  Stockfish is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  Stockfish is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+// Definition of layer AffineTransformSparseInput of NNUE evaluation function
+
+#ifndef NNUE_LAYERS_AFFINE_TRANSFORM_SPARSE_INPUT_H_INCLUDED
+#define NNUE_LAYERS_AFFINE_TRANSFORM_SPARSE_INPUT_H_INCLUDED
+
+#include <iostream>
+#include <algorithm>
+#include <array>
+#include <type_traits>
+#include "../nnue_common.h"
+#include "affine_transform.h"
+#include "simd.h"
+
+/*
+  This file contains the definition for a fully connected layer (aka affine transform) with block sparse input.
+*/
+
+namespace Stockfish::Eval::NNUE::Layers {
+
+#if defined(USE_SSSE3)
+  const std::array<std::array<std::uint16_t, 8>, 256> lookup_indices = [](){
+      std::array<std::array<std::uint16_t, 8>, 256> v{};
+      for (int i = 0; i < 256; ++i)
+      {
+          int j = i;
+          int k = 0;
+          while(j)
+          {
+              const IndexType lsbIndex = __builtin_ctzl(std::uint32_t(j));
+              j &= j - 1;
+              v[i][k] = lsbIndex;
+              ++k;
+          }
+      }
+      return v;
+  }();
+  const std::array<unsigned, 256> lookup_count = [](){
+      std::array<unsigned, 256> v;
+      for (int i = 0; i < 256; ++i)
+      {
+          int j = i;
+          int k = 0;
+          while(j)
+          {
+              j &= j - 1;
+              ++k;
+          }
+          v[i] = k;
+      }
+      return v;
+  }();
+
+  template<const IndexType InputDimensions>
+  void find_nnz(const std::int32_t* input, std::uint16_t* out, IndexType& count_out) {
```

Comment:
This function could use some comments, as it's not obvious what it does from the name. It gives back indices to non-zero inputs, but, crucially, it groups them in groups of 4, because later accumulation does 4 outputs at a time. So the first 4 inputs have index 0, the next 4 have index 1, and so on. Input chunk is considered non-zero if any of the 4 inputs in the group is non-zero.